sapp/ui/trace.py

```diff
@@ -403,3 +438,94 @@ def trace_kind_to_shared_text_kind(trace_kind: Optional[TraceKind]) -> SharedTex
         return SharedTextKind.SINK
 
     raise AssertionError(f"{trace_kind} is invalid")
+
+
+def to_sarif(
+    session: Session, issue: IssueQueryResult, tool: str, output_features: bool = False
+) -> SARIFCodeflowsObject:
+    postcondition_initial_frames = initial_frames(
+        session,
+        issue.issue_instance_id,
+        TraceKind.POSTCONDITION,
+    )
+    precondition_initial_frames = initial_frames(
+        session,
+        issue.issue_instance_id,
+        TraceKind.PRECONDITION,
+    )
+    postcondition_navigation = navigate_trace_frames(
+        session,
+        postcondition_initial_frames,
+        set(issue.source_kinds),
+        set(issue.sink_kinds),
+    )
+    precondition_navigation = navigate_trace_frames(
+        session,
+        precondition_initial_frames,
+        set(issue.source_kinds),
+        set(issue.sink_kinds),
+    )
+    trace_tuples = _create_trace_tuples(
+        reversed(postcondition_navigation)
+    ) + _create_trace_tuples(precondition_navigation)
+    codeflows = [{"threadFlows": [{"locations": []}]}]
+    nesting_level = 0
+    for t in trace_tuples:
+        location = _sarif_codeflow_location_from_trace_tuple(
+            t.trace_frame, tool, nesting_level, True
+        )
+        codeflows[0]["threadFlows"][0]["locations"].append(location)
+        nesting_level += 1
+
+    return codeflows
+
+
+def _create_trace_tuples(
+    navigation: Iterable[Tuple[TraceFrameQueryResult, int]]
+) -> List[TraceTuple]:
+    return [
+        TraceTuple(
+            trace_frame=trace_frame,
+            branches=branches,
+            missing=trace_frame.caller == "",
+        )
+        for trace_frame, branches in navigation
+    ]
+
+
+def _sarif_codeflow_location_from_trace_tuple(
+    trace_frame: TraceFrameQueryResult,
+    tool: str,
+    nesting_level: int = 1,
+    output_features: bool = False,
+) -> Dict[str, Union[SARIFCodeflowLocationObject, Dict[str, str]]]:
+    features_str = titos = ""
+    if output_features:
+        frame_features = [
+            text.contents
+            for text in trace_frame.shared_texts
+            if text.kind is SharedTextKind.FEATURE
+        ]
+        if frame_features:
+            features_str = f"features: {frame_features}"
+        if trace_frame.titos and len(trace_frame.titos.split(";")) > 0:
+            titos = f"via {len(trace_frame.titos.split(';'))} propagators"
```

Comment:
This is pre-existing but it looks like we store a list as a string separated by `;`. We should really just have the list within `trace_frame`.

Also I'm not sure what those `titos` are compared to `features`.
Finally, "propagators" sounds weird to me but no strong feelings :p

Reply:
> Finally, "propagators" sounds weird to me but no strong feelings :p

happy if you have better naming (remember I am terrible with names :D). I will as ChatGTP for help :D

Reply:
> Also I'm not sure what those titos are compared to features.

I am not comparing I am just trying to give a summary for each frame through the number of titos and list of features available in the frame.